inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp

```diff
@@ -3,27 +3,280 @@
 //
 
 #include "mkldnn_input_node.h"
+#include "common/cpu_memcpy.h"
 #include "mkldnn_extension_utils.h"
+
 #include <string>
 #include <tuple>
 #include <algorithm>
 #include <utils/general_utils.h>
 #include <ngraph/ops.hpp>
+#include <ie_parallel.hpp>
 #include <ie_ngraph_utils.hpp>
 #include <blob_factory.hpp>
 #include "caseless.hpp"
 #include "common/cpu_memcpy.h"
 #include "common/cpu_convert.h"
 #include "utils/cpu_utils.hpp"
+#include <cpu/x64/jit_generator.hpp>
 
 using namespace mkldnn;
 using namespace MKLDNNPlugin;
 using namespace InferenceEngine;
 using namespace details;
 using namespace ngraph::op;
+using namespace dnnl::impl::cpu::x64;
+using namespace Xbyak;
+
+namespace {
+
+struct jit_has_subnormals : public jit_generator {
+    DECLARE_CPU_JIT_AUX_FUNCTIONS(jit_has_subnormals)
+
+    typedef struct {
+        const float* src;
+        const size_t count;
+        bool hasSubnormals;
+    } args_t;
+
+    typedef void (*fn_t)(const args_t*);
+
+    jit_has_subnormals() {
+        jit_ker_ = nullptr;
+    }
+
+    fn_t get() {
+        return jit_ker() || create_kernel() == dnnl::impl::status::success
+                ? (fn_t)jit_ker()
+                : nullptr;
+    }
+
+protected:
+    void foreach(const Xbyak::Reg64& idx,
+                 size_t step,
+                 const Xbyak::Reg64& end,
+                 std::function<void(const Xbyak::Reg64&)> && fn) {
+        Label loop, exit;
+
+        L(loop);
+        cmp(idx, end);
+        jge(exit);
+
+        fn(idx);
+
+        add(idx, step);
+        jmp(loop);
+        L(exit);
+    }
+
+    void copy_floats(const Xbyak::Reg64& dst,
+                     const Xbyak::Reg64& src,
+                     const Xbyak::Reg64& size) {
+        push(rsi);
+        push(r15);
+
+        xor_(rsi, rsi);
+
+        foreach(rsi, 1, size, [&, this](const Xbyak::Reg64& idx) {
+            mov(r15d, dword[src + idx * sizeof(float)]);
+            mov(dword[dst + idx * sizeof(float)], r15d);
+        });
+
+        pop(r15);
+        pop(rsi);
+    }
+};
+
+struct jit_has_subnormals_avx2 : public jit_has_subnormals {
```

Comment:
Why you need separate function for sse? They are almost the same. Could you please handle copypaste by using template? We have a lot of good examples in plugin which use this approach.

Reply:
fixed